0xPolygon/polygon-edge
# Add blockRangeLimit and batchRequestLimit JSON-RPC flags to help prevent node DDoS

# Description

This PR introduces 2 new server flags that help prevent node DDoS:
`--json-rpc-block-range-limit` and `--json-rpc-batch-request-limit`

### Block Range Limit

During testing, we observed that a couple JSON-RPC requests that include block ranges could allow an attacker to use up all the resource on a particular node if they specify large ranges, essentially DDoS-ing that node. For example, the following request could take down any polygon-edge node running on a chain that consists of many blocks:
```
{"jsonrpc":"2.0","id":1,"method":"eth_getLogs","params":[{"fromBlock":"0","toBlock":"latest"}]}
```

To prevent such an attack, this PR enforces a default block range limit of 1000, optionally giving the node operator the ability to override this by specifying a value for `--json-rpc-block-range-limit` on their node's server command.

### Batch Request Limit

In addition to the above, since we currently don't have a limit for the allowed size of a batch JSON-RPC request body, attackers could spam RPC nodes with requests that contain thousands of batched requests, for example:
```
[
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 0}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 1}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 2}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 3}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 4}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 5}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 6}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 7}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 8}, 
	{"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 9},
        ...
        {"jsonrpc": "2.0", "method": "eth_getBlockByNumber", "params": ["latest", true], "id": 99999},
]
```

Part of this PR also enforces a default batch request size of 20, optionally giving the node operator the ability to override this by specifying a value for `--json-rpc-batch-request-limit` on their node's server command.


# Changes include


# Checklist


## Testing


### Manual tests

1. Run this branch on a node
2. Attempt to query with `{"jsonrpc":"2.0","id":1,"method":"eth_getLogs","params":[{"fromBlock":"0","toBlock":"1002"}]}` (this exceeds the default value of 1000). Observe error message in response.
3. Attempt to query with batch JSON-RPC request containing 25 entries (this exceeds the default of 20). Observe error message in response.
4. Add `--json-rpc-block-range-limit 1100` and `--json-rpc-batch-request-limit 30` to the server command on the same node and restart the node.
5. Attempt both queries from step 2. and 3. and observe that the node now responds with success responses.

# Documentation update

Will create a separate documentation PR once this PR is approved.
